pintracker/optracker/operationtracker.go

```diff
@@ -159,3 +221,64 @@ func (opt *OperationTracker) GetOpContext(c *cid.Cid) context.Context {
 	}
 	return op.Context()
 }
+
+// Filter returns a slice of api.PinInfos that had associated
+// Operations that matched the provided filter. Note, only supports
+// filters of type OperationType or Phase, any other type
+// will result in a nil slice being returned.
+func (opt *OperationTracker) Filter(filters ...interface{}) []api.PinInfo {
+	var pinfos []api.PinInfo
+	opt.mu.RLock()
+	defer opt.mu.RUnlock()
+	ops := filterOpsMap(opt.operations, filters)
+	for _, op := range ops {
+		pinfos = append(pinfos, opt.unsafePinInfo(op))
+	}
+	return pinfos
+}
+
+// FilterOps returns a slice that only contains operations
+// with the matching filter. Note, only supports
+// filters of type OperationType or Phase, any other type
+// will result in a nil slice being returned.
+func (opt *OperationTracker) FilterOps(filters ...interface{}) []*Operation {
+	var fltops []*Operation
+	opt.mu.RLock()
+	defer opt.mu.RUnlock()
+	for _, op := range filterOpsMap(opt.operations, filters) {
+		fltops = append(fltops, op)
+	}
+	return fltops
+}
+
+func filterOpsMap(ops map[string]*Operation, filters []interface{}) map[string]*Operation {
```

Comment:
This seems unefficient. It loops the whole map for every filter. Instead of looping it once and check, for every element, if it matches one of the filters. I'm assuming that looping the map is generally more expensive than looping the filters, even if the final number of comparisons is the same.

I also don't see the advantage in using recursion here when a simple `for` over the filters solves the issue. It should be significantly cheaper than creating a whole new context to call the function, but mostly is about readability (there are not so many filter types).

Reply:
> This seems unefficient. It loops the whole map for every filter. 

It may be inefficient but it isn't looping the whole map for every filter. It only loops the whole map once, and then only loops the filtered map for consecutive filters. Reading my godoc of the filter-related functions after not looking at them for a while, I realise they don't convey this at all so I shall change those. 

> Instead of looping it once and check, for every element, if it matches one of the filters. 

This wouldn't allow for the usage seen in `stateless.SyncAll` where the map is filtered by an operation, i.e. Pin/Unpin, and then by phase, i.e. Error, essentially this function results in an [intersection](https://en.wikipedia.org/wiki/Intersection_(set_theory)) of the provided filters on the operation map. 

> I also don't see the advantage in using recursion here when a simple for over the filters solves the issue. 

The reason why I went with recursion as it resulted in much more readable code and code climate complained about my initial non-recursive implementation.

Reply:
Ah, I was reading this as an OR filter, not an AND. If the problem is filtering by operations and then by phase, we might as well use TrackerStatus as filter value directly

Reply:
The issue with using a TrackerStatus is that then you can't get all `errors` or all `pinOps` regardless of phase.

Reply:
I got halfway through implementing it to use TrackerStatus when I realised this.

Reply:
Use `-1` for all ?

Reply:
But that would not be filtering anything?

Reply:
Sorry, misread. Maybe making filter be a slice of slices. [][]TrackerStatus were `{{status1, status2}, {status3}}` means (Status1 OR status 2) AND (status3) ? Or simply two methods (filterOR, filterAND)

Reply:
thoughts on this as a way of doing the filtering via trackerstatuses: https://play.golang.org/p/ykGXNsnderW

Reply:
So I implemented this and it adds a lot of complexity that really outweighs the benefits of getting rid of the `interface{}` parameter. And it is actually more error prone because the conversion from `TrackerStatus` to `OperationType` and `Phase` isn't perfect. On top of that, the behaviour is less intuitive. 

I am of the opinion that we leave it as is and if we have issues with this later, we tackle it as a separate thing to the stateless tracker. I am happy to push up the changes as separate PR?